source/blog/_posts/2017-11-17-on-testing-strategies.md

```diff
@@ -0,0 +1,282 @@
+
+So now lets take a look at our `MyServiceTest` the way it looks like in integration testing:
+
+<img src="/blog/img/advantages-of-integration-testing/my-service-integration-test.svg" />
+
+This test still covers all n-paths of the class `MyService`! It covers most paths in it's directly dependant classes `MyRepository` and `ValidationService` and covers a few paths in their dependencies, and so on. You can imagine this like different saturations of a color:
+
+<img src="/blog/img/advantages-of-integration-testing/my-service-integration-test-with-saturation.svg" />
+
+Now we no longer make up the output of any subsidiary service. We even noticed that using the real storage has virtually no impact on the test performance, but creates a huge benefit by testing the friction between these different services in our test suite at all times [(see factor X)](https://12factor.net/dev-prod-parity).
+
+The big picture then looks like this:
+
+<img src="/blog/img/advantages-of-integration-testing/standard-application-with-diamond.svg" />
+
+The mocks in this case come from a specially set up database and this is the only real downside to this approach. Your fixtures may diverge from a production setting! We have seen this in the past and usually got real bugs from this. But even in these cases the problem arose from a single point of failure which is far better then mocking in every single place.
+
+**Good:**
+
+* Helps during development, creates entry points for every possible error
+* Almost as fast as unit tests
+* The real data is used
+* Creates an architecture of well integrated parts
+
+**Bad:**
+
+* Almost doubles the amount of code in your project
+* Fixture data will never show 100% production state
+* Will also not give you a 100% certainty that everything works
+
+**Verdict:** We us it!
+
+## A word on TDD
+
+I personally am a huge fan of TDD and practice it almost always when developing or exploring software. This practice is not lost here. Just because you have to implement stuff that you are using does not reduce the amount of freedom you have in the 30 second cycle of test driven development. On the contrary - at least for me - it tends to create more realistic assumptions on the scope of classes. By doing almost only integration testing all my tests look and feel the same. The usual friction that occurs when combining multiple unit tested classes is by far reduced because the integration assumptions all - by definition -  hold up against the real implementations.
```

Comment:
ggf. I am personally

Reply:
double white space between - and hold up against the real im

Reply:
google sagt geht auch so rum